c7n/resources/s3.py

```diff
@@ -2494,3 +2494,95 @@ def process_bucket(self, bucket):
 
         s3.put_bucket_lifecycle_configuration(
             Bucket=bucket['Name'], LifecycleConfiguration={'Rules': config})
+
+
+@actions.register('set-bucket-encryption')
+class SetBucketEncryption(BucketActionBase):
+    """Action enables default encryption on S3 buckets
+
+    `enabled`: boolean Optional: Defaults to True
+    `crypto`: aws:kms | AES256` Optional: Defaults to AES256
+    `key`: arn, or id kms key
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: s3-enable-default-encryption-kms
+                resource: s3
+                actions:
+                  - type: set-bucket-encryption
+                    crypto: aws:kms
+                    key: 1234abcd-12ab-34cd-56ef-1234567890ab
+
+              - name: s3-enable-default-encryption-aes256
+                resource: s3
+                actions:
+                  - type: set-bucket-encryption
+                    enabled: false
+    """
+
+    schema = {
+        'type': 'object',
+        'additionalProperties': False,
+        'properties': {
+            'type': {'enum': ['set-bucket-encryption']},
+            'enabled': {'type': 'boolean'},
+            'crypto': {'enum': ['aws:kms', 'AES256']},
+            'key': {'type': 'string'}
+        },
+        'dependencies': {
+            'key': {
+                'properties': {
+                    'crypto': {'pattern': 'aws:kms'}
+                },
+                'required': ['crypto']
+            }
+        }
+    }
+
+    permissions = ('s3:PutEncryptionConfiguration',
+                's3:GetEncryptionConfiguration')
+
+    def process(self, buckets):
+        with self.executor_factory(max_workers=3) as w:
+            results = []
+
+            if ('enabled' in self.data and self.data['enabled']) or 'enabled' not in self.data:
+                config = {'Rules': [
+                    {
+                        'ApplyServerSideEncryptionByDefault': {
+                            'SSEAlgorithm': 'AES256'
+                        }
+                    }
+                ]}
+                if 'key' in self.data and self.data['crypto'] == 'aws:kms':
+                    config['Rules'][0]['ApplyServerSideEncryptionByDefault']['SSEAlgorithm'] = 'aws:kms'
+                    config['Rules'][0]['ApplyServerSideEncryptionByDefault']['KMSMasterKeyID'] = self.data['key']
+                elif 'key' in self.data and not self.data['crypto'] == 'aws:kms':
+                    logging.warning('Warning: \'key\' not valid for non KMS encryption, using AES256')
+
+            elif 'enabled' in self.data and not self.data['enabled']:
+                config = None
+
+            futures = {w.submit(self.process_bucket, b, config) : b for b in buckets}
```

Comment:
If action does not specify enabled, this will throw a NameError since config is only defined in the if or elif sections.

Reply:
I don't believe this would result in a `NameError` the following scenarios should all be handled as follows:

`'enabled': True`
- `('enabled' in self.data and self.data['enabled']) or 'enabled' not in self.data` first block in the `or` evaluates to `True`

`'enabled': False`
- `'enabled' in self.data and not self.data['enabled']` evaluates to `True`

`'enabled': None`
- `('enabled' in self.data and self.data['enabled']) or 'enabled' not in self.data` second block in the `or` evaluates to `True`

`'enabled': SomeNonBooleanValue`
- Not valid per schema

Reply:
I see that now.